### DISCUSSION TITLE
Getting Your data directory is invalid every other day.

### DISCUSSION THREAD

**Discussion by apoorv569** [nextcloud/all-in-one#3462] [Questions] [answered]
I have been using `Nextcloud` `AIO` in a `VM` on my `Proxmox` server for well over a year. It worked great, but recently I have been getting this error message almost every day,

![2023-09-30_08-23](https://github.com/nextcloud/all-in-one/assets/53279454/bdd3de45-7275-4a61-a57a-32d1d48a35a4)

`df -h` says the `NFS` mount is `Stale file handle`
```
apoorv@nextcloud:~$ df -h
df: /mnt/data: Stale file handle
Filesystem      Size  Used Avail Use% Mounted on
tmpfs           692M  6.8M  685M   1% /run
/dev/sda2        50G   32G   15G  69% /
tmpfs           3.2G     0  3.2G   0% /dev/shm
tmpfs           5.0M     0  5.0M   0% /run/lock
tmpfs           642M  4.0K  642M   1% /run/user/1000
```

I have been using `NFS` share for all my `VM`s for as long as I'm using `Proxmox` which is almost a year now. All other `VM`s seems to be doing fine.

Not sure why this `VM` break every day. I didn't change any `NFS` option its the same I have been using always.

If I restart the VM it says,
```
mount.nfs: mounting 10.10.10.231:/export/nextcloud failed, reason given by server: No such file or directory
```

However `showmount -e 10.10.10.231` show the mount exists and the IP is correct as well.

`10.10.10.231` is my `NAS` BTW running `OpenMediaVault`.

To fix this I either have to restart the whole server or restart `NFS` service on my `NAS` which breaks all connections forcing me to restart all `VM`s anyway.

**Comment by szaimen** [comment, marked as answer]
Hi, this is definitely not caused by AIO so I would recommend to ask in a OpenMediaVault forum for help on this

**Reply by apoorv569** [reply]
I see. Also when I do restart NFS service and mount the directory properly and Nextcloud starts, 90% of times I can't login, I enter my username password and I just get re-directed back to login page again. But when I do manage to login I get "Failed to get SOME_WIDGET's info" and also "You are not logged in" even though refreshing the page does open up dashboard.

I can reinstall `Nextcloud-AIO` from scratch if need be, but I wanna know if I just give it the current data directory will it repopulate all the stuff as in user's files and all? I can create users account again no problem, there's only 4-5 people on the server so no issue there.

I just don't want to upload their files all over again, as there is a lot of data 100+ GB.

I wanna fix this before something else happens and data gets corrupted or something. Its been more than a week now I haven't been able to use nextcloud without fighting with it first.

**Reply by szaimen** [reply]
> I can reinstall `Nextcloud-AIO` from scratch if need be, but I wanna know if I just give it the current data directory will it repopulate all the stuff as in user's files and all?

If you use the built-in backup solution and then import the backup archive on a new clean AIO instance then yes.

**Reply by apoorv569** [reply]
I don't, but I can enable it. What does it backs up BTW? Does it backup data as well?

**Reply by szaimen** [reply]
> Does it backup data as well?

It backs up all nextcloud related data

**Reply by apoorv569** [reply]
I see. I don't have enough space to back up all the users data, can I exclude it?

**Reply by szaimen** [reply]
> can I exclude it?

No

**Reply by apoorv569** [reply]
Is there any way to repair the installation?

**Reply by szaimen** [reply]
Yes, fix the VM

**Reply by apoorv569** [reply]
I would, if I could find the issue as it seems nothing to be wrong with the VM.

**EDIT**: Another thing I remember before this started happening is that I `git pull`ed the `aio` repo and ran `docker compose pull` manually rather than using `aio` interface one time, could that have broke something?

**Reply by szaimen** [reply]
> Another thing I remember before this started happening is that I `git pull`ed the `aio` repo and ran `docker compose pull` manually rather than using `aio` interface one time, could that have broke something?

No not that I am aware of

**Reply by apoorv569** [reply]
I see. Well I'm still investigating, will let you know if I become aware of something.

**Reply by apoorv569** [reply]
@szaimen OK, so I tested for a week now. I revert to the last `git` commit I had by running `git reset --hard HEAD@{1}` and updated the VM, it seem to fix the `stale file handle` issue for now.

But I sometimes still face this issue, where it keeps saying I'm not logged in and failed to fetch stuff etc, sometimes it works fine there is no particular timing or thing I do that makes it work..
![2023-10-16_12-48](https://github.com/nextcloud/all-in-one/assets/53279454/b8aff3b6-cd95-465c-852b-871909ff5f16)

When I get these errors, neither CPU usage is high nor RAM and neither the `NFS` mount is broken in some way, all seems good but not sure why these errors show up and refreshing the page more than 2 or 3 times logs me out and then it takes multiple tries to log back in.

**Reply by apoorv569** [reply]
@szaimen  OK, so I received an update today for all containers and a new container`docker-socket-proxy` was added. ATM everything is working (until its not :disappointed:).

I saw some errors from `nextcloud` container logs during update though,
```
Turned off maintenance mode
Resetting log level
Nextcloud 27.1.2
Posting notifications to users that are admins...
Posting 'Nextcloud update to 27.1.2.1 successful!' to: admin
Done!
Restoring the status of apps. This can take a while...
Enabling deck...
Config value enabled for app deck set to Your data directory is invalid.
Ensure there is a file called ".ocdata" in the root of the data directory.

An unhandled exception has been thrown:
Exception: Environment not properly prepared. in /var/www/html/lib/private/Console/Application.php:166

ace:
#0 /var/www/html/console.php(99): OC\Console\Application->loadCommands(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#1 /var/www/html/occ(11): require_once('/var/www/html/c...')

}
Enabling integration_mastodon...
Enabling drawio...
Config value enabled for app drawio set to Your data directory is invalid.
Ensure there is a file called ".ocdata" in the root of the data directory.

An unhandled exception has been thrown:
Exception: Environment not properly prepared. in /var/www/html/lib/private/Console/Application.php:166

ace:
#0 /var/www/html/console.php(99): OC\Console\Application->loadCommands(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#1 /var/www/html/occ(11): require_once('/var/www/html/c...')

}
Enabling integration_github...
Enabling notes...
Config value enabled for app notes set to Your data directory is invalid.
Ensure there is a file called ".ocdata" in the root of the data directory.

An unhandled exception has been thrown:
Exception: Environment not properly prepared. in /var/www/html/lib/private/Console/Application.php:166

ace:
#0 /var/www/html/console.php(99): OC\Console\Application->loadCommands(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#1 /var/www/html/occ(11): require_once('/var/www/html/c...')

}
Enabling calendar...
Config value enabled for app calendar set to Your data directory is invalid.
Ensure there is a file called ".ocdata" in the root of the data directory.

An unhandled exception has been thrown:
Exception: Environment not properly prepared. in /var/www/html/lib/private/Console/Application.php:166

ace:
#0 /var/www/html/console.php(99): OC\Console\Application->loadCommands(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#1 /var/www/html/occ(11): require_once('/var/www/html/c...')

}
Enabling notify_push...
Config value enabled for app notify_push set to Your data directory is invalid.
Ensure there is a file called ".ocdata" in the root of the data directory.

An unhandled exception has been thrown:
Exception: Environment not properly prepared. in /var/www/html/lib/private/Console/Application.php:166

ace:
#0 /var/www/html/console.php(99): OC\Console\Application->loadCommands(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#1 /var/www/html/occ(11): require_once('/var/www/html/c...')

}
Enabling files_archive...
Enabling richdocuments...
Enabling tasks...
Config value enabled for app tasks set to Your data directory is invalid.
Ensure there is a file called ".ocdata" in the root of the data directory.

An unhandled exception has been thrown:
Exception: Environment not properly prepared. in /var/www/html/lib/private/Console/Application.php:166

ace:
#0 /var/www/html/console.php(99): OC\Console\Application->loadCommands(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#1 /var/www/html/occ(11): require_once('/var/www/html/c...')

}
Enabling extract...
Enabling end_to_end_encryption...
Enabling integration_google...
Enabling health...
Config value enabled for app health set to Your data directory is invalid.
Ensure there is a file called ".ocdata" in the root of the data directory.

An unhandled exception has been thrown:
Exception: Environment not properly prepared. in /var/www/html/lib/private/Console/Application.php:166

ace:
#0 /var/www/html/console.php(99): OC\Console\Application->loadCommands(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#1 /var/www/html/occ(11): require_once('/var/www/html/c...')

}
Enabling spreed...
Enabling contacts...
Config value enabled for app contacts set to Your data directory is invalid.
Ensure there is a file called ".ocdata" in the root of the data directory.

An unhandled exception has been thrown:
Exception: Environment not properly prepared. in /var/www/html/lib/private/Console/Application.php:166

ace:
#0 /var/www/html/console.php(99): OC\Console\Application->loadCommands(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#1 /var/www/html/occ(11): require_once('/var/www/html/c...')

}
Enabling approval...
Enabling forms...
Enabling integration_gitlab...
Config value enabled for app integration_gitlab set to Your data directory is invalid.
Ensure there is a file called ".ocdata" in the root of the data directory.

An unhandled exception has been thrown:
Exception: Environment not properly prepared. in /var/www/html/lib/private/Console/Application.php:166

ace:
#0 /var/www/html/console.php(99): OC\Console\Application->loadCommands(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#1 /var/www/html/occ(11): require_once('/var/www/html/c...')

}
Enabling files_antivirus...
Enabling mail...
Enabling files_fulltextsearch...
Enabling fulltextsearch_elasticsearch...
Enabling maps...
Enabling twofactor_totp...
Config value enabled for app twofactor_totp set to Your data directory is invalid.
Ensure there is a file called ".ocdata" in the root of the data directory.

An unhandled exception has been thrown:
Exception: Environment not properly prepared. in /var/www/html/lib/private/Console/Application.php:166

ace:
#0 /var/www/html/console.php(99): OC\Console\Application->loadCommands(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#1 /var/www/html/occ(11): require_once('/var/www/html/c...')

}
Enabling news...
Enabling announcementcenter...
Enabling fulltextsearch...
Enabling integration_reddit...
Enabling files_fulltextsearch_tesseract...
Enabling electronicsignatures...
Config value enabled for app electronicsignatures set to Your data directory is invalid.
Ensure there is a file called ".ocdata" in the root of the data directory.

An unhandled exception has been thrown:
Exception: Environment not properly prepared. in /var/www/html/lib/private/Console/Application.php:166

ace:
#0 /var/www/html/console.php(99): OC\Console\Application->loadCommands(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#1 /var/www/html/occ(11): require_once('/var/www/html/c...')

}
Doing some optimizations...
 - Repair MySQL collation
     - Not a mysql database -> nothing to do
 - Repair mime types
 - Clean tags and favorites
     - 0 tags of deleted users have been removed.
     - 0 tags for delete files have been removed.
     - 0 tag entries for deleted tags have been removed.
     - 0 tags with no entries have been removed.
 - Repair invalid shares
 - Move .step file of updater to backup location
 - Add move avatar background job
 ```
 
 after this everything was good and the server started fine..
 
 This,
 ```
 Ensure there is a file called ".ocdata" in the root of the data directory.
 ```
 
 I'm not sure why it says that, because (from Nextcloud VM) the `NFS` share is mounted correct and working, 
![2023-10-17_15-14](https://github.com/nextcloud/all-in-one/assets/53279454/a3c13e97-516e-4dc5-b6e0-16dd31dd29e2)

The file `.ocdata` exists, but it shows as 0 bytes. Is this a problem?

**Reply by szaimen** [reply]
I still have the feeling that your NFS mount is unreliable. Otherwise I cannot explain this error message as I never experienced this with block storage.

**Comment by SN4K3D** [comment]
Hi, @apoorv569 today i have the same issue with nfs mount in my kubernetes cluster with nextcloud data folder , the mount point is ok but , every day the container loose Stale file handle of the data folder. 
When i restart the container , everythings is okay , but is not a good practice. 

i have a mover on my nas with mount bonding and the data of day are moved on other disk ( cache -> Array ) i suppose the issue is here , maybe inode file are not the same.

Have you find a trick for you'r issue ?
